src/compiler/checker.ts

```diff
@@ -22966,19 +23218,44 @@ namespace ts {
                       case 'boolean': break
                   }
 
-                  In the first clause (case `number` and `string`) the
-                  implied type is number | string.
+                  In the first clause (case `number` and `string`) the implied type is number | string.
 
                   In the default clause we de not compute an implied type.
 
-                  In the third clause (case `number` and `boolean`)
-                  the naive implied type is number | boolean, however
-                  we use the type facts to narrow the implied type to
-                  boolean. We know that number cannot be selected
+                  In the third clause (case `number` and `boolean`) the naive implied type is number | boolean, however
+                  we use the type facts to narrow the implied type to boolean. We know that number cannot be selected
                   because it is caught in the first clause.
                 */
+                const expr = (<TypeOfExpression>switchStatement.expression).expression;
                 const impliedType = getTypeWithFacts(getUnionType(clauseWitnesses.map(text => getImpliedTypeFromTypeofGuard(type, text) || type)), switchFacts);
-                return getTypeWithFacts(mapType(type, narrowUnionMemberByTypeof(impliedType)), switchFacts);
+                if (isMatchingReference(reference, expr)) {
+                    if (hasDefaultClause) {
+                        return filterType(type, t => (getTypeFacts(t) & switchFacts) === switchFacts);
+                    }
+                    return getTypeWithFacts(mapType(type, narrowUnionMemberByTypeof(impliedType)), switchFacts);
+                }
+                const impliedUndefined = !(switchFacts & 0b111111 && !(switchFacts & TypeFacts.EQUndefined) || (switchFacts & TypeFacts.NEUndefined));

```

Comment:
I haven’t looked closely at this function yet but no way are we keeping the literal `0b111111` in here 😆

Reply:
This might be a name problem. I could not find a good name for the union, would use `TypeFacts.TypeofEQString|TypeFacts.TypeofEQNumber|.....` to replace for now.

Reply:
Simplify this line by using `hasDefaultClause`. no such wired literal now!